caikit/runtime/work_management/abortable_context.py

```diff
@@ -0,0 +1,178 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# Standard
+from queue import SimpleQueue
+from typing import Dict
+import abc
+import ctypes
+import threading
+import uuid
+
+# First Party
+import alog
+
+# Local
+from caikit.runtime.types.aborted_exception import AbortedException
+
+log = alog.use_channel("ABORT-ACTION")
+
+
+class AbortableContextBase(abc.ABC):
+    @abc.abstractmethod
+    def abort(self):
+        """Called to abort work in progress"""
+
+
+class ActionAborter(abc.ABC):
+    """Simple interface to wrap up a notification that an action must abort.
+
+    Children of this class can bind to any notification tool (e.g. grpc context)
+    """
+
+    @abc.abstractmethod
+    def must_abort(self) -> bool:
+        """Indicate whether or not the action must be aborted"""
+
+    @abc.abstractmethod
+    def set_context(self, context: AbortableContextBase):
+        """Set the abortable context that must be notified to abort work"""
+
+    @abc.abstractmethod
+    def unset_context(self):
+        """Unset any abortable context already held. Do not notify it that work should abort"""
+
+
+class ThreadInterrupter:
+    """This class implements a listener which will observe all ongoing work in `AbortableContexts`
+     and raise exceptions in the working threads if they need to be aborted.
+
+    This offers a performance advantage over the old `AbortableActions`, since only one extra
+    listener thread is created that lives for the whole lifetime of the program. The caveat
+    is that all work to be aborted must be running in a thread that is safe to kill: this may
+    not be safe to use with asyncio tasks running in an event loop.
+    """
+
+    def __init__(self):
+        # Using a SimpleQueue because we don't need the Queue's task api
+        self._queue = SimpleQueue()
+        self._thread = None
+
+        self._context_thread_map: Dict[uuid.UUID, int] = {}
+
+        self._shutdown = True
+
+    def start(self):
+        if not self._shutdown:
+            log.debug("ThreadInterrupter already started")
+            return
+        log.debug("Starting ThreadInterrupter")
+        self._shutdown = False
+        self._thread = threading.Thread(target=self._watch_loop)
+        self._thread.start()
+
+    def stop(self):
+        if self._shutdown:
+            log.debug("ThreadInterrupter already shut down")
+            return
+
+        log.info("Stopping ThreadInterrupter")
+        self._shutdown = True
+        self._queue.put(0)
```

Comment:
Why do we need to put this sentinel `0` on the queue? Is this to avoid getting stuck in `_watch_loop`'s call to `self._queue.get()`? This feels like a special sentinel value that should be defined as `self.__class__._SHUTDOWN` or something, but I don't think it's actually ever used on the receiving end. Either way, a comment explaining this `self._queue.put(0)` would be nice.